v2/pkg/engine/resolve/resolve.go

```diff
@@ -306,7 +306,12 @@ func (s *sub) startWorkerWithHeartbeat() {
 
 	for {
 		select {
-		case <-s.resolver.ctx.Done(): // Skip sending events if the resolver is shutting down
+		case <-s.ctx.ctx.Done():
+			// Complete when the client request context is done for synchronous subscriptions
+			s.complete()
```

Comment:
Chan is closed in .complete and used to coordinate the shutdown of synchronous subscription.